File: api/core/hunt/ingestion/tools/dns.py
"""DNS resolution + record enumeration.

Pure stdlib socket — no upstream HTTP, no API key, always works.
The output goes through the Evidence Vault so the resolver's answer
is hash-attested, even though no third party is involved.
"""
from __future__ import annotations

import json
import socket
from typing import Any, List, Mapping

import httpx

from hunt.ingestion.base import Finding, ToolFunction
from hunt.models import EntityKind, SourceTool


class DnsTool(ToolFunction):
    name = SourceTool.DNS
    accepts = {"domain"}
    emits = {EntityKind.IPV4, EntityKind.IPV6, EntityKind.SUBDOMAIN}
    description = "Resolves A/AAAA records via the system resolver."
    per_request_timeout = 5.0

    def _fetch(self, query: Mapping[str, Any]) -> httpx.Response:
        target = query["target"].strip().lower()
        records: dict = {"target": target, "records": {}}
        try:
            infos = socket.getaddrinfo(target, None)
            v4, v6 = set(), set()
            for fam, *_rest, sockaddr in infos:
                ip = sockaddr[0]
                if fam == socket.AF_INET:
                    v4.add(ip)
                elif fam == socket.AF_INET6:
                    v6.add(ip)
            records["records"]["A"] = sorted(v4)
            records["records"]["AAAA"] = sorted(v6)
        except socket.gaierror as exc:
            records["error"] = str(exc)
        body = json.dumps(records).encode("utf-8")
        return httpx.Response(
            200, content=body, headers={"content-type": "application/json"}
        )

    def _parse(self, raw: bytes, query: Mapping[str, Any]) -> List[Finding]:
        records = json.loads(raw.decode("utf-8"))
        findings: List[Finding] = []
        host = records.get("target") or query["target"]
        for ip in records.get("records", {}).get("A", []):
            findings.append(
                Finding(
                    source_tool=self.name,
                    entity_kind=EntityKind.IPV4,
                    entity_value=ip,
                    attributes={"record_type": "A", "host": host},
                )
            )
        for ip in records.get("records", {}).get("AAAA", []):
            findings.append(
                Finding(
                    source_tool=self.name,
                    entity_kind=EntityKind.IPV6,
                    entity_value=ip,
                    attributes={"record_type": "AAAA", "host": host},
                )
            )
        return findings

File: lib/tools/dns.ts
/**
 * DNS resolution + record enumeration.
 *
 * Pure stdlib resolver — no upstream HTTP, no API key, always works.
 * The output goes through the Evidence Vault so the resolver's
 * answer is hash-attested.
 */

import { promises as dns } from "node:dns";
import { ToolFunction, type FetchResult, type Finding } from "./base";

export class DnsTool extends ToolFunction {
  readonly name = "dns";
  readonly accepts = new Set(["domain", "subdomain", "person", "org"]);
  readonly emits = new Set(["ipv4", "ipv6", "subdomain"]);
  readonly description = "Resolves A/AAAA records via the system resolver.";
  readonly per_request_timeout_ms = 5_000;

  protected async _fetch(query: { target: string; kind?: string | null }): Promise<FetchResult> {
    const target = (query.target || "").trim().toLowerCase();
    const v4 = new Set<string>();
    const v6 = new Set<string>();
    let err: string | null = null;
    try {
      const a = await dns.resolve4(target).catch(() => []);
      const aaaa = await dns.resolve6(target).catch(() => []);
      for (const r of a) v4.add(r);
      for (const r of aaaa) v6.add(r);
    } catch (e) {
      err = String(e);
    }
    const body = Buffer.from(
      JSON.stringify({
        target,
        records: { A: [...v4].sort(), AAAA: [...v6].sort() },
        ...(err ? { error: err } : {}),
      })
    );
    return { bytes: body, contentType: "application/json" };
  }

  protected _parse(raw: Buffer, query: { target: string }): Finding[] {
    const data = JSON.parse(raw.toString("utf-8")) as {
      target?: string;
      records?: { A?: string[]; AAAA?: string[] };
    };
    const findings: Finding[] = [];
    const host = data.target || query.target;
    for (const ip of data.records?.A ?? []) {
      findings.push({
        source_tool: this.name,
        entity_kind: "ipv4",
        entity_value: ip,
        attributes: { record_type: "A", host },
      });
    }
    for (const ip of data.records?.AAAA ?? []) {
      findings.push({
        source_tool: this.name,
        entity_kind: "ipv6",
        entity_value: ip,
        attributes: { record_type: "AAAA", host },
      });
    }
    return findings;
  }
}
